### PR TITLE
feat: implement router api.

### PR DESCRIPTION
**Pull request by mchlumsky**
Implements the router api as per: https://docs.openstack.org/api-ref/network/v2/#routers-routers

### CODE REVIEW

**Comment by dtantsur** [comment]
Ouch, I'm not sure how I missed this PR :( Sorry! I'll try to get to it as soon as possible, feel free to nag me next time.

File: src/network/api.rs [hunk 1]
```diff
@@ -14,16 +14,103 @@
...
+    Ok(())
+}
+
+/// Remove extra routes from a router.
+pub fn remove_extraroutes<S>(session: &Session, id: S, routes: Vec<Route>) -> Result<()>
+where
+    S: AsRef<str>,
+{
+    trace!(
+        "Rmove extra routes {:?} from router {}",
```
Comment (dtantsur) [inline src/network/api.rs:53]: nit: "Remove"

File: src/network/protocol.rs [hunk 2]
```diff
@@ -414,6 +435,149 @@ pub struct PortsRoot {
...
+    pub network_id: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub enable_snat: Option<bool>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub external_fixed_ips: Option<Vec<FixedIp>>,
+}
+
+/// A route.
+#[derive(Debug, Clone, Deserialize, Serialize, Copy, PartialEq)]
+pub struct Route {
```
Comment (dtantsur) [inline src/network/protocol.rs:458]: There is an identical `HostRoute` structure which I'd prefer to be reused.

**Review by dtantsur — APPROVED** [review]
I've left a few comments. Since this PR has been hanging here for a while, I intend to fix them myself in a follow-up, but first I'd like to hear from you if you have any objections.

**Comment by mchlumsky** [comment]
No worries about the PR response time. I figured we're all busy with our day jobs! ;)

File: examples/create-router-with-external-gateway.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    let name = env::args().nth(1).expect("Provide a router name");
+    let external_network = env::args()
+        .nth(2)
+        .expect("Provide an external network name or ID");
+
+    let external_network = os
+        .get_network(&external_network)
+        .unwrap_or_else(|_| panic!("Network {} not found", &external_network.to_string()));
+
+    let external_gateway_info = ExternalGatewayInfo {
```
[the remark this answers is not in the document]
Comment (mchlumsky) [reply examples/create-router-with-external-gateway.rs:39]: So you may have noticed I inspired myself from other existing API structures (the Network API especially) and there were some similar structures to ExternalGatewayInfo like AllocationPool and HostRoute that were part of structures like Subnet that have their own API endpoints. I emulated that to stay consistent but that doesn't mean we need to continue doing that by any means. 
So the new() constructor sounds ok but can we make sure to allow setting the other fields somehow? Maybe keep the struct public and implement the Default trait?

File: src/network/routers.rs [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+    pub fn remove_router_interface(
+        &mut self,
+        subnet_id: Option<&String>,
+        port_id: Option<&String>,
+    ) -> Result<()> {
+        api::remove_router_interface(&self.session, self.id(), subnet_id, port_id)
+    }
+
+    /// Add route to router.
+    pub fn add_extraroutes(&mut self, routes: Vec<protocol::Route>) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (mchlumsky) [reply src/network/routers.rs:248]: Agreed. It looks a little weird and lopsided.

File: src/network/protocol.rs [hunk 5]
```diff
@@ -414,6 +435,149 @@ pub struct PortsRoot {
     pub ports: Vec<Port>,
 }
 
+/// ConntrackHelper.
+#[derive(Debug, Clone, Deserialize, PartialEq)]
+pub struct ConntrackHelper {
```
[the remark this answers is not in the document]
Comment (mchlumsky) [reply src/network/protocol.rs:440]: I'm actually not that familiar with conntrack helpers and I just implemented the API as per the apiref document.
I'd love to see some doc too! ;)
So +1 for adding some doc for this struct.

File: src/network/protocol.rs [hunk 6]
```diff
@@ -414,6 +435,149 @@ pub struct PortsRoot {
...
+    pub helper: String,
+    pub protocol: String,
+    pub port: u16,
+}
+
+/// External gateway information.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct ExternalGatewayInfo {
+    pub network_id: String,
```
[the remark this answers is not in the document]
Comment (mchlumsky) [reply src/network/protocol.rs:449]: I see there is an IdOrName enum in the osproto crate. Perhaps we can use that but we'll need to add code to do the actually query to the neutron API for the conversion.

File: src/network/routers.rs [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    update_field! {
+        #[doc = "Set the administrative state of the router."]
+        set_admin_state_up, with_admin_state_up -> admin_state_up: bool
+    }
+
+    transparent_property! {
+        #[doc = "Availability zone candidates for the router"]
+        availability_zone_hints: ref Option<Vec<String>>
```
[the remark this answers is not in the document]
Comment (mchlumsky) [reply src/network/routers.rs:81]: Yeah, the router API always returns a list even if there are no availability_zone_hints.
+1

File: src/network/api.rs (hunk 1 above)
Comment (mchlumsky) [reply src/network/api.rs:53]: +1

File: src/network/protocol.rs (hunk 2 above)
Comment (mchlumsky) [reply src/network/protocol.rs:458]: Perhaps we can make Route a type alias to HostRoute or vice versa.

File: examples/create-router-with-external-gateway.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dtantsur) [reply examples/create-router-with-external-gateway.rs:39]: The existing API structures are not, unfortunately, perfect examples, and I tend to make decisions on case-by-case basis. With this comment I'm following two goals:

1) Make constructing `ExternalGatewayInfo` a bit less cumbersome. `Default` helps, a separate `new` for the simplest case helps as well. I'd actually do both.

2) Prevent breakages if `ExternalGatewayInfo` is ever extended. Now, many structures here a *hopefully* stable (e.g. I assume `HostRoutes` won't change). I have no insight into whether this one will ever expand. Marking all fields as private is a bit nuclear, we can instead use the new `#[non_exhaustive]` attribute to explicitly state our intention to extend this structure in the future.

File: src/network/protocol.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dtantsur) [reply src/network/protocol.rs:449]: There is a bit of ugly magic with `*Ref*` types and this trait: https://github.com/dtantsur/rust-openstack/blob/master/src/common/types.rs#L27. It's super confusing, so don't waste too many brain cells on it if you don't feel like it.

If you do feel like it, the `*Ref` types track whether they have been created from an instance (so the ID is validated) or a raw string value (so the ID may need validation and conversion). We could put a `pub network: NetworkRef` here and then validate it before creating/saving a route. The downside: we can no longer use the same structure for high-level API and for the protocol level.

File: src/network/protocol.rs (hunk 2 above)
Comment (dtantsur) [reply src/network/protocol.rs:458]: Yep, and deprecate `HostRoute`.

**Comment by dtantsur** [comment]
I've merged the PR as it is. Feel free to follow-up on any of the comments, otherwise I'll get to them before the next release (which may take me some time since $daily_jobs as you rightfully noticed).

**Pull request merged**